sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java

```diff
@@ -216,53 +195,49 @@ void readBatch(int total, WritableColumnVector column) throws IOException {
           boolean needTransform = castLongToInt || isUnsignedInt32 || isUnsignedInt64;
           column.setDictionary(new ParquetDictionary(dictionary, needTransform));
         } else {
-          updater.decodeDictionaryIds(num, rowId, column, dictionaryIds, dictionary);
+          updater.decodeDictionaryIds(readState.offset - startOffset, startOffset, column,
+            dictionaryIds, dictionary);
         }
       } else {
-        if (column.hasDictionary() && rowId != 0) {
+        if (column.hasDictionary() && readState.offset != 0) {
           // This batch already has dictionary encoded values but this new page is not. The batch
           // does not support a mix of dictionary and not so we will decode the dictionary.
-          updater.decodeDictionaryIds(rowId, 0, column, dictionaryIds, dictionary);
+          updater.decodeDictionaryIds(readState.offset, 0, column, dictionaryIds, dictionary);
         }
         column.setDictionary(null);
         VectorizedValuesReader valuesReader = (VectorizedValuesReader) dataColumn;
-        defColumn.readBatch(num, rowId, column, maxDefLevel, valuesReader, updater);
+        defColumn.readBatch(readState, column, valuesReader, updater);
       }
-
-      valuesRead += num;
-      rowId += num;
-      total -= num;
     }
   }
 
-  private void readPage() {
+  private int readPage() {
     DataPage page = pageReader.readPage();
-    // TODO: Why is this a visitor?
-    page.accept(new DataPage.Visitor<Void>() {
+    return page.accept(new DataPage.Visitor<Integer>() {
       @Override
-      public Void visit(DataPageV1 dataPageV1) {
+      public Integer visit(DataPageV1 dataPageV1) {
```

Comment:
To avoid boxing, we can assign the valueCount to a local variale and return it, e.g.
```
int pageValueCount = 0;
page.accept(new DataPage.Visitor<Integer>() {
  ...
  pageValueCount = readPageV1(dataPageV1);
  return null;
  ...
}
return pageValueCount;
```

Reply:
I'm not sure you can do that, since the variable has to be final in order to be accessed within inner class.

Also it probably doesn't matter that much since this is not the hot path (I'm very tempted to remove the visitor pattern here though...)

Reply:
ah I see, let's leave it then.